src/tests/utests/xi_utest_sft_logic.c

```diff
@@ -231,6 +231,292 @@ XI_TT_TESTCASE_WITH_SETUP(
+                .list_len = 3,
+                .list     = three_file_list}};
+
+        int32_t download_order[] = {-1, -1, 2};
+        xi_sft_context_t sft_context;
+        sft_context.updateable_files_download_order = download_order;
+        sft_context.update_current_file             = NULL;
+        sft_context.update_message_fua              = &file_update_available;
+
+        xi_state_t state = _xi_sft_select_next_resource_to_download( &sft_context );
+        tt_want_int_op( XI_STATE_OK, ==, state );
+        tt_want_ptr_op( sft_context.update_current_file, ==, &three_file_list[2] );
+    } )
+
+XI_TT_TESTCASE_WITH_SETUP(
+    xi_utest__select_next_resource_to_download__all_negative_index,
+    xi_utest_setup_basic,
+    xi_utest_teardown_basic,
+    NULL,
+    {
+        // ARRANGE
+        xi_control_message_file_desc_ext_t three_file_list[3];
+        memset( three_file_list, 0, 3 * sizeof( xi_control_message_file_desc_ext_t ) );
+
+        xi_control_message_t file_update_available = {
+            .file_update_available = {
+                .common = {.msgtype = XI_CONTROL_MESSAGE_SC__SFT_FILE_UPDATE_AVAILABLE,
+                           .msgver  = 1},
+                .list_len = 3,
+                .list     = three_file_list}};
+
+        int32_t download_order[] = {-1, -1, -2};
+        xi_sft_context_t sft_context;
+        sft_context.updateable_files_download_order = download_order;
+        sft_context.update_current_file             = NULL;
+        sft_context.update_message_fua              = &file_update_available;
+
+        xi_state_t state = _xi_sft_select_next_resource_to_download( &sft_context );
+        tt_want_int_op( XI_STATE_OK, ==, state );
+        tt_want_ptr_op( NULL, ==, sft_context.update_current_file );
+    } )
+
+XI_TT_TESTCASE_WITH_SETUP(
+    xi_utest__select_next_resource_to_download__duplicate_index_ok,
+    xi_utest_setup_basic,
+    xi_utest_teardown_basic,
+    NULL,
+    {
+        // ARRANGE
+        xi_control_message_file_desc_ext_t three_file_list[3];
+        memset( three_file_list, 0, 3 * sizeof( xi_control_message_file_desc_ext_t ) );
+
+        xi_control_message_t file_update_available = {
+            .file_update_available = {
+                .common = {.msgtype = XI_CONTROL_MESSAGE_SC__SFT_FILE_UPDATE_AVAILABLE,
+                           .msgver  = 1},
+                .list_len = 3,
+                .list     = three_file_list}};
+
+        int32_t download_order[] = {1, 1, 1};
+        xi_sft_context_t sft_context;
+        sft_context.updateable_files_download_order = download_order;
+        sft_context.update_current_file             = NULL;
+        sft_context.update_message_fua              = &file_update_available;
+
+        xi_state_t state = _xi_sft_select_next_resource_to_download( &sft_context );
+        tt_want_int_op( XI_STATE_OK, ==, state );
+        tt_want_ptr_op( sft_context.update_current_file, ==, &three_file_list[1] );
```

Comment:
hm, why is the second selected?

Reply:
why, not first or last?

Reply:
because the first index in the download order array points to the second file in the file list.  This ensures that we iteratively go left to right in the list and find the first file that's available.  Even if the index is repeated, as this function doesn't guarantee index uniqueness.

Reply:
ah, right